File: 1312-minimum-insertion-steps-to-make-a-string-palindrome/1312-minimum-insertion-steps-to-make-a-string-palindrome.cpp
class Solution {
public:
    // vector<vector<int>>dp;
    // int recur(string s,int i,int j){
    //     if(i>=j)return 0;
    //     if(dp[i][j]!=-1)return dp[i][j];
    //     return dp[i][j]=s[i]==s[j]?recur(s,i+1,j-1):1+min(recur(s,i+1,j),recur(s,i,j-1));
    // }
    
    int minInsertions(string s) {
        vector<int> memo(s.length(),0);
        int prev,temp;
        for(int i=s.length()-2;i>=0;i--)
        {
            prev=0;                                       
            for(int j=i;j<s.length();j++)
            {
                temp=memo[j];                             
                memo[j]=s[i]==s[j]?prev:1+min(memo[j],memo[j-1]);
                prev=temp;                        
            }
        }
        return memo[s.length()-1];
    }
};

File: 1312-minimum-insertion-steps-to-make-a-string-palindrome/1312-minimum-insertion-steps-to-make-a-string-palindrome.java
class Solution {
//     int dp[][];
//     public int memo(String s,int i,int j){
//         if(i>=j)return 0;
    
//         if(dp[i][j]!=-1)return dp[i][j];
        
//         if(s.charAt(i)==s.charAt(j))return dp[i][j]= memo(s,i+1,j-1);
        
//         return dp[i][j]= 1+Math.min(memo(s,i+1,j),memo(s,i,j-1));
//     }
    
    int [][]dp;
    public int lps(String s,int i,int j){
        if(i==j)return 1;
        if(i>j)return 0;
        
        if(dp[i][j]!=-1)return dp[i][j];
        if(s.charAt(i)==s.charAt(j))return dp[i][j]=2+lps(s,i+1,j-1);
        return dp[i][j]=Math.max(lps(s,i+1,j),lps(s,i,j-1));
    }
    
    public int minInsertions(String s) {
        int n=s.length();
        dp=new int[n][n];
        for(int []i:dp)
            Arrays.fill(i,-1);
        return n-lps(s,0,n-1);
    }
}